### ISSUE TITLE
[scala]  src/c_api/c_api_symbolic.cc:385: InferShapeKeyword argument name label not found.

### ISSUE THREAD

**Issue by benqua** [dmlc/mxnet#4755]
For bugs or installation issues, please provide the following information.
The more information you provide, the more likely people will be able to help you.

## Environment info
Operating System:
Linux Ubuntu

MXNet version:
latest from source.

## Error Message:
```
 src/c_api/c_api_symbolic.cc:385: InferShapeKeyword argument name label not found.
Candidate arguments:
	[0]data
	[1]conv1_1_weight
	[2]conv1_1_bias
	[3]conv1_2_weight
	[4]conv1_2_bias
	[5]conv2_1_weight
	[6]conv2_1_bias
	[7]conv2_2_weight
	[8]conv2_2_bias
	[9]conv3_1_weight
	[10]conv3_1_bias
	[11]conv3_2_weight
	[12]conv3_2_bias
	[13]conv3_3_weight
	[14]conv3_3_bias
	[15]conv4_1_weight
	[16]conv4_1_bias
	[17]conv4_2_weight
	[18]conv4_2_bias
	[19]conv4_3_weight
	[20]conv4_3_bias
	[21]conv5_1_weight
	[22]conv5_1_bias
	[23]conv5_2_weight
	[24]conv5_2_bias
	[25]conv5_3_weight
	[26]conv5_3_bias
	[27]fc6_weight
	[28]fc6_bias
	[29]fc7_weight
	[30]fc7_bias
	[31]fc8_weight
	[32]fc8_bias
	[33]softmax_label

Stack trace returned 5 entries:
[bt] (0) /tmp/mxnet3809050951967264516/mxnet-scala(_ZN4dmlc15LogMessageFatalD1Ev+0x39) [0x7fbb32fd5c99]
[bt] (1) /tmp/mxnet3809050951967264516/mxnet-scala(_ZN5mxnet14MatchArgumentsIN4nnvm6TShapeEEEvRKNS1_12IndexedGraphERKSt13unordered_mapISsT_St4hashISsESt8equal_toISsESaISt4pairIKSsS7_EEEPSt6vectorIS7_SaIS7_EEPKc+0x639) [0x7fbb33bf63c9]
[bt] (2) /tmp/mxnet3809050951967264516/mxnet-scala(MXSymbolInferShape+0x15cb) [0x7fbb33beeccb]
[bt] (3) /tmp/mxnet3809050951967264516/mxnet-scala(Java_ml_dmlc_mxnet_LibInfo_mxSymbolInferShape+0x2a4) [0x7fbb32fcdc84]
[bt] (4) [0x7fbbd4017774]

[error] (run-main-0) java.lang.NegativeArraySizeException
java.lang.NegativeArraySizeException
	at ml.dmlc.mxnet.LibInfo.mxSymbolInferShape(Native Method)
	at ml.dmlc.mxnet.Symbol.inferShape(Symbol.scala:243)
	at ml.dmlc.mxnet.Symbol.inferShape(Symbol.scala:233)
	at ml.dmlc.mxnet.FeedForward.initParams(FeedForward.scala:112)
	at ml.dmlc.mxnet.FeedForward.initSymbolParams(FeedForward.scala:313)
	at ml.dmlc.mxnet.FeedForward.fit(FeedForward.scala:248)
	at ml.dmlc.mxnet.FeedForward.fit(FeedForward.scala:259)
	at ml.dmlc.mxnet.FeedForward.fit(FeedForward.scala:265)
	at ml.dmlc.mxnet.FeedForward.fit(FeedForward.scala:270)
	at name.benq.VGG16FineTuned0$.main(VGG16FineTuned0.scala:70)
	at name.benq.VGG16FineTuned0.main(VGG16FineTuned0.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)

```
## Minimum reproducible example
I load and modify a VGG16 model.
The last layer is:

``` scala

val softmax = Symbol.SoftmaxOutput(name="softmax")()(Map("data" -> fc8))

```

Then I initialize a FeefForward network and train it:
``` scala

    val vgg16 = new FeedForward(
      symbol = softmax,
      ctx = Array(Context.gpu()),
      numEpoch = 3,
      epochSize = 1,
      optimizer = new SGD(0.01f),
      initializer = new Xavier(),
      batchSize = 32,
      argParams = newArgParams,
      auxParams = newAuxParams,
      allowExtraParams = false,
      beginEpoch = 0
    )

    vgg16.fit(trainIter, valIter, new Accuracy())

```
My trainIter shape is 3022,3,204,204 and its data shape is: data -> (32,3,204,204) label -> (32,8)

From #1263 I understand that the name of the last layer should match the one from the iterator.
However, 
 - I see no way to set the name of the labels in the NDArrayIter API
 - I cannot change the name of the last softmax layer to "label" because _label is automatically added to the name. (So, if I name it "label", it ends in "label_label").

How can I make these two names match?  
Isn't the "name" parameter missing from the NDArrayiter API?

**Comment by javelinjs** [comment]
How did you use the NDArrayIter?

**Comment by benqua** [comment]
Use or build?

I use it in the vgg16.fit function above.

**Comment by javelinjs** [comment]
Oh, I see the problem. In `NDArrayIter.scala`

```scala
val initData: IndexedSeq[(String, NDArray)] = IO.initData(_dataList, false, "data")
val initLabel: IndexedSeq[(String, NDArray)] = IO.initData(_labelList, true, "label")
```

always give label name as 'label'. Could you help to add `dataName` and `labelName` arguments to `NDArrayIter` ?

**Comment by benqua** [comment]
Yes, you're right!

Sure, I can and will. Not sure when exactly, but in the next few days.

By the way, is there a way to force the name of the output of the SoftmaxOutput layer to "label" (which would be the other way to solve the issue)?

**Comment by Ldpe2G** [comment]
@benqua try 
```scala
val softmax = Symbol.SoftmaxOutput()()(Map("data" -> fc8))
```

**Comment by benqua** [comment]
@javelinjs I did a PR to add the names of the dataset as parameter to the NDArrayIter constructor.

@Ldpe2G It works this way, thanks!

Not sure if I should open another issue for this, but when I try to fit my FeedForward  that ends with
``` scala
   val numClasses = 8
(...)
    val fc8 = Symbol.FullyConnected("fc8")()(Map("data" -> drop7, "num_hidden" -> numClasses))
    val label = Symbol.Variable("label")
    val softmax = Symbol.SoftmaxOutput()()(Map("data" -> fc8, "label" -> label))

```
with a categorical ouptut encoded with NDArray.onehotEncode, I get a Shape mismatch error:
`[error] (run-main-0) java.lang.IllegalArgumentException: requirement failed: label (32,8) and prediction (32)should have the same length.`
However, it seems correct that the Shape of label is (32,8) as the batchSize is 32 and the number of oneHotEncode output (= number of classes) is 8.
With the above network, why is the Shape of output of the FeedForward (32) instead of (32,8)?
Is it a bug, should I open a new issue or did I miss something?

**Comment by Ldpe2G** [comment]
@benqua the label shape should be just (batchSize), you don't need to do one hot yourself.
Also the error message should come from EvalMetric.scala.
